scripts/e2e_setup_cluster.sh

```diff
@@ -51,23 +53,38 @@ generate_k8_api_data() {
 
 create_cluster() {
   [ -z "${mount_dir}" ] && echo "### no mount directory set" && exit 1
+
+  # create list of worker nodes
+  workers="$(for i in "${KIND_WORKER_NAMES[@]}"; do echo "  - role: worker"; done)"
+
+  # create KinD configuration file
+  exec 3<> "${PWD}"/kindConfig.yaml
+
+    # Let's print Kind configuration to file to fd 3
+    echo "kind: Cluster" >&3
+    echo "apiVersion: kind.x-k8s.io/v1alpha4" >&3
+    echo "featureGates:" >&3
+    echo "  DownwardAPIHugePages: true" >&3
+    echo "kubeadmConfigPatches:" >&3
+    echo "- |" >&3
+    echo "  kind: ClusterConfiguration" >&3
+    echo "  apiServer:" >&3
+    echo "    extraArgs:" >&3
+    echo "      admission-control-config-file: /etc/kubernetes/pki/ac.yaml" >&3
+    echo "nodes:" >&3
+    echo "  - role: control-plane" >&3
+    echo "    extraMounts:" >&3
+    echo "    - hostPath: \"${mount_dir:?}\"" >&3
+    echo "      containerPath: \"/etc/kubernetes/pki\"" >&3
+    echo "${workers}" >&3
+
+  # Close fd 3
+  exec 3>&-
+
   # deploy cluster with kind
-  cat <<EOF | kind create cluster --config=-
-kind: Cluster
-apiVersion: kind.x-k8s.io/v1alpha4
-kubeadmConfigPatches:
-- |
-  kind: ClusterConfiguration
-  apiServer:
-    extraArgs:
-      admission-control-config-file: /etc/kubernetes/pki/ac.yaml
-nodes:
-  - role: control-plane
-    extraMounts:
-    - hostPath: "${mount_dir:?}"
-      containerPath: "/etc/kubernetes/pki"
-  - role: worker
-EOF
+  kind create cluster --config="${PWD}"/kindConfig.yaml
```

Comment:
can you add the retry function before kind create ...
Ive seen sometimes kind create fails for no reason in like 1-15 deployments. I suggest:
```
retry kind delete cluster && kind create cluster ...
```
youll have to move retry func above this func I think so its visible for calling.

Reply:
Never experience such behavior of KinD, but I can do such modification.